src/open_producten/producttypen/management/commands/load_upl.py

```diff
@@ -1,68 +1,137 @@
+            if file:
+                created_count, update_count, removed_count = self._parse_csv_file(file)
+            else:
+                created_count, update_count, removed_count = self._parse_csv_url(url)
+        except CommandError:
+            raise
         except Exception as e:
-            raise CommandError(str(e))
+            raise CommandError(f"Something went wrong: {str(e)}")
+        self.stdout.write(
+            "Done\n"
+            f"Created {created_count} product names.\n"
+            f"Updated {update_count} product names.\n"
+            f"{removed_count} product names did not exist in the csv."
+        )
+
+    def _parse_csv_file(self, file: str):
+        _check_if_csv_extension(file)
+
+        with open(file, encoding="utf-8-sig") as f:
+            data = csv.DictReader(f)
+            return self._load_upl(data)
+
+    def _parse_csv_url(self, url: str):
+        _check_if_csv_extension(url)
 
-        self.stdout.write(f"Done ({created_count} objects).")
+        response = requests.get(url)
+        if response.status_code != 200:
+            raise CommandError(f"Url returned status code: {response.status_code}.")
 
-    def load_upl(self, data: list[UniformProductName]) -> int:
-        count = 0
+        content = StringIO(response.text)
+        data = csv.DictReader(content)
+        return self._load_upl(data)
+
+    @transaction.atomic
+    def _load_upl(self, data: csv.DictReader) -> tuple[int, int, int]:
+        created_count = 0
+        updated_count = 0
         upn_updated_list = []
+        columns = {
+            "uri": "URI",
+            "name": "UniformeProductnaam",
+        }
+
+        if missing_columns := [
+            f"'{key}'" for key in columns.values() if key not in data.fieldnames
+        ]:
+            raise CommandError(
+                f"Column(s) {', '.join(missing_columns)} do not exist in the CSV."
+            )
+
+        for i, row in enumerate(data):
+            uri = row[columns["uri"]]
+            name = row[columns["name"]]
+
+            if not name or not uri:
+                self.stdout.write(
+                    f"Skipping index {i} because of missing column(s) {' or '.join(columns.values())}."
+                )
+                continue
 
-        for obj in data:
             upn, created = UniformProductNaamModel.objects.update_or_create(
-                uri=obj.uri,
-                defaults={"naam": obj.name, "is_verwijderd": False},
+                uri=uri,
+                defaults={"naam": name, "is_verwijderd": False},
             )
             upn_updated_list.append(upn.id)
 
             if created:
-                count += 1
+                created_count += 1
+            else:
+                updated_count += 1
 
-        UniformProductNaamModel.objects.exclude(id__in=upn_updated_list).update(
+        removed_count = UniformProductNaamModel.objects.exclude(
```

Comment:
This gets the count but IDK how. Should it not return a queryset? or is str(querset) always the count?

Reply:
`.update()` returns the number of affected rows.
[docs](https://docs.djangoproject.com/en/4.2/ref/models/querysets/#django.db.models.query.QuerySet.update)